qiskit/algorithms/minimum_eigen_solvers/qaoa/qaoa_circuit.py

```diff
@@ -0,0 +1,209 @@
+from qiskit.circuit import Parameter, ParameterVector
+from qiskit.circuit.library import BlueprintCircuit
+from qiskit.opflow import OperatorBase, I, X, CircuitStateFn, H, CircuitOp, EvolutionFactory
+
+
+class QAOACircuit(BlueprintCircuit):
+    """Global X phases and parameterized problem hamiltonian."""
+
+    def __init__(self,
+                 cost_operator: OperatorBase,
+                 param_p: int,
+                 initial_state: Optional[QuantumCircuit] = None,
+                 mixer_operator: Optional[Union[QuantumCircuit, OperatorBase]] = None,
+                 name: str = "qaoa"):
+        """
+        Constructor, following the QAOA paper https://arxiv.org/abs/1411.4028
+
+        Args:
+            cost_operator: The operator representing the cost of the optimization problem,
+                denoted as U(B, gamma) in the original paper.
+            param_p: The integer parameter p, which determines the depth of the circuit,
+                as specified in the original paper.
+            initial_state: An optional initial state to use.
+            mixer_operator: An optional custom mixer to use instead of the global X-rotations,
+                denoted as U(B, beta) in the original paper. Can be an operator or
+                an optionally parameterized quantum circuit.
+            name: A name of the circuit, default 'qaoa'
+        Raises:
+            TypeError: invalid input
+        """
+        super().__init__(name=name)
+        self._cost_operator = cost_operator
+        self._param_p = param_p
+        self._initial_state = initial_state
+        self._mixer_operator = mixer_operator
+
+        # set this circuit as a not-built circuit
+        self._num_qubits = 0
+        self._num_parameters = 0
+        self._bounds = None
+        self._mixer = None
+        self._data = None
+
+    def _check_configuration(self, raise_on_failure: bool = True) -> bool:
+        valid = True
+
+        if self._cost_operator is None:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The operator representing the cost of "
+                                     "the optimization problem is not set")
+        if self._param_p is None or self._param_p < 1:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The integer parameter p, which determines the depth "
+                                     "of the circuit, either not set or set to non-positive value")
+
+        return valid
+
+    def _build(self) -> None:
+        """Build the circuit."""
+        if self._data:
+            return
+
+        self._check_configuration()
+        self._data = []
+
+        # calculate bounds, num_parameters, mixer
+        self._calculate_parameters()
+
+        # parametrize circuit and build it
+        param_vector = ParameterVector("θ", self._num_parameters)
+        circuit = self._construct_circuit(param_vector)
```

Comment:
Why is this in a separate method? For readability? 🙂

Reply:
Historical reasons and readability. I prefer short methods and shallow indentation in general.